File: Chagok/frontend/src/lib/api/client-portal.ts
/**
 * Client Portal API Functions
 * 003-role-based-ui Feature - US4
 */

import { apiRequest } from './client';
import type {
  ClientDashboardResponse,
  ClientCaseListResponse,
  ClientCaseDetailResponse,
  EvidenceUploadRequest,
  EvidenceUploadResponse,
  EvidenceConfirmResponse,
} from '@/types/client-portal';

/**
 * Get client dashboard data
 */
export async function getClientDashboard() {
  return apiRequest<ClientDashboardResponse>('/client/dashboard', {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Get client's case list
 */
export async function getClientCases() {
  return apiRequest<ClientCaseListResponse>('/client/cases', {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Get client case detail
 */
export async function getClientCaseDetail(caseId: string) {
  return apiRequest<ClientCaseDetailResponse>(`/client/cases/${caseId}`, {
    method: 'GET',
    credentials: 'include',
  });
}

/**
 * Request presigned URL for evidence upload
 */
export async function requestEvidenceUpload(caseId: string, data: EvidenceUploadRequest) {
  return apiRequest<EvidenceUploadResponse>(`/client/cases/${caseId}/evidence`, {
    method: 'POST',
    credentials: 'include',
    body: JSON.stringify(data),
  });
}

/**
 * Confirm evidence upload completion
 */
export async function confirmEvidenceUpload(caseId: string, evidenceId: string, uploaded: boolean) {
  return apiRequest<EvidenceConfirmResponse>(
    `/client/cases/${caseId}/evidence/${evidenceId}/confirm`,
    {
      method: 'POST',
      credentials: 'include',
      body: JSON.stringify({ uploaded }),
    }
  );
}

/**
 * Upload file to S3 using presigned URL
 */
export async function uploadFileToS3(
  uploadUrl: string,
  file: File,
  onProgress?: (percent: number) => void
): Promise<{ success: boolean; error?: string }> {
  return new Promise((resolve) => {
    const xhr = new XMLHttpRequest();

    xhr.upload.addEventListener('progress', (event) => {
      if (event.lengthComputable && onProgress) {
        const percent = Math.round((event.loaded / event.total) * 100);
        onProgress(percent);
      }
    });

    xhr.addEventListener('load', () => {
      if (xhr.status >= 200 && xhr.status < 300) {
        resolve({ success: true });
      } else {
        resolve({ success: false, error: `Upload failed: ${xhr.status}` });
      }
    });

    xhr.addEventListener('error', () => {
      resolve({ success: false, error: 'Network error during upload' });
    });

    xhr.open('PUT', uploadUrl);
    xhr.setRequestHeader('Content-Type', file.type);
    xhr.send(file);
  });
}

File: Chagok/backend/app/api/client_portal.py
"""
Client Portal API Router
003-role-based-ui Feature - US4 (T061-T066)

API endpoints for client portal including dashboard, case viewing, and evidence upload.
"""

import logging

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.dependencies import get_db, require_role, get_file_storage_service
from app.core.error_messages import ErrorMessages
from app.services.client_portal_service import ClientPortalService
from app.domain.ports.file_storage_port import FileStoragePort
from app.schemas.client_portal import (
    ClientDashboardResponse,
    ClientCaseListResponse,
    ClientCaseDetailResponse,
    EvidenceUploadRequest,
    EvidenceUploadResponse,
    EvidenceConfirmRequest,
    EvidenceConfirmResponse,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/client", tags=["client-portal"])


# ============== T062: Dashboard Endpoint ==============

@router.get("/dashboard", response_model=ClientDashboardResponse)
async def get_client_dashboard(
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["client"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Get client dashboard data.

    Returns:
        - User name and greeting
        - Current case summary with progress
        - Progress steps visualization
        - Assigned lawyer contact info
        - Recent activities
        - Unread message count
    """
    service = ClientPortalService(db, file_storage_port=file_storage_port)
    try:
        return service.get_dashboard(user_id)
    except ValueError as e:
        logger.warning(f"Client dashboard not found for user {user_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CLIENT_DASHBOARD_NOT_FOUND
        )


# ============== T063: Case List Endpoint ==============

@router.get("/cases", response_model=ClientCaseListResponse)
async def get_client_cases(
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["client"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Get list of client's cases.

    Returns list of cases where the client is a member.
    """
    service = ClientPortalService(db, file_storage_port=file_storage_port)
    return service.get_case_list(user_id)


# ============== T064: Case Detail Endpoint ==============

@router.get("/cases/{case_id}", response_model=ClientCaseDetailResponse)
async def get_client_case_detail(
    case_id: str,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["client"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Get detailed case information.

    Args:
        case_id: Case ID to retrieve

    Returns:
        Detailed case information including:
        - Case metadata and status
        - Progress steps
        - Lawyer info
        - Evidence list
        - Recent activities
    """
    service = ClientPortalService(db, file_storage_port=file_storage_port)
    try:
        return service.get_case_detail(user_id, case_id)
    except PermissionError:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="사건에 접근할 권한이 없습니다"
        )
    except ValueError as e:
        logger.warning(f"Case detail not found for user {user_id}, case {case_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=ErrorMessages.CASE_NOT_FOUND
        )


# ============== T065: Evidence Upload Endpoint ==============

@router.post("/cases/{case_id}/evidence", response_model=EvidenceUploadResponse)
async def request_evidence_upload(
    case_id: str,
    request: EvidenceUploadRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["client"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Request presigned URL for evidence upload.

    Args:
        case_id: Case ID to upload evidence to
        request: Upload request with file metadata

    Returns:
        Evidence ID and presigned S3 upload URL

    Note:
        - File size limit: 100MB
        - URL expires in 5 minutes
        - Evidence is marked as 'pending' until confirmed
    """
    # Validate file size (100MB limit)
    max_size = 100 * 1024 * 1024  # 100MB
    if request.file_size > max_size:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"File size exceeds limit of {max_size // (1024*1024)}MB"
        )

    service = ClientPortalService(db, file_storage_port=file_storage_port)
    try:
        response = service.request_evidence_upload(
            user_id=user_id,
            case_id=case_id,
            file_name=request.file_name,
            file_type=request.file_type,
            file_size=request.file_size,
            description=request.description,
        )

        # T066: Audit logging for evidence upload
        service.log_evidence_upload(
            user_id=user_id,
            case_id=case_id,
            evidence_id=response.evidence_id,
            file_name=request.file_name,
        )

        return response

    except PermissionError:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="사건에 접근할 권한이 없습니다"
        )
    except ValueError as e:
        logger.warning(f"Evidence upload request failed for case {case_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="증거 업로드 요청에 실패했습니다. 입력값을 확인해주세요"
        )


@router.post("/cases/{case_id}/evidence/{evidence_id}/confirm", response_model=EvidenceConfirmResponse)
async def confirm_evidence_upload(
    case_id: str,
    evidence_id: str,
    request: EvidenceConfirmRequest,
    db: Session = Depends(get_db),
    user_id: str = Depends(require_role(["client"])),
    file_storage_port: FileStoragePort = Depends(get_file_storage_service),
):
    """
    Confirm evidence upload completion.

    Call this after successfully uploading to S3 to mark
    the evidence as ready for processing.
    """
    service = ClientPortalService(db, file_storage_port=file_storage_port)

    try:
        result = service.confirm_evidence_upload(
            user_id=user_id,
            case_id=case_id,
            evidence_id=evidence_id,
            uploaded=request.uploaded,
        )
        return EvidenceConfirmResponse(**result)
    except ValueError as e:
        error_msg = str(e).lower()
        if "not found" in error_msg:
            logger.warning(f"Evidence not found for confirmation: {evidence_id}: {e}")
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=ErrorMessages.EVIDENCE_NOT_FOUND
            )
        logger.warning(f"Evidence confirmation failed: {evidence_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="증거 확인에 실패했습니다"
        )
    except PermissionError as e:
        logger.warning(f"Permission denied for evidence confirmation: {evidence_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="증거 확인 권한이 없습니다"
        )
